tests/identity/test_slurmrestd_identity.py

```diff
@@ -148,44 +149,69 @@ def test_acquire_token__gets_a_token_from_slurm_if_one_is_not_in_the_cache(
     token_path = mock_slurmrestd_api_cache_dir / "dummy-user.token"
     assert not token_path.exists()
 
-    process_mock = mock.Mock()
-    process_mock.communicate.return_value = (
-        bytes("SLURM_JWT=dummy-token\n".encode("utf-8")),
-        bytes(str().encode("utf-8")),
+    now = datetime.now()
+    username = "dummy-user"
+
+    mocked_datetime.now = mock.Mock()
```

Comment:
Oh, this can be a fraught approach! I recommend using a library designed for this such as:
* https://pypi.org/project/time-machine/
* https://github.com/spulec/freezegun

Reply:
what do you mean by fraught approach? The code won't run in freezed time so I think freezing the time will apply an unreal behaviour

Reply:
I mean that relying on a mock to manipulate the current time is dangerous. There are lots of edge cases and odd behavior.

When testing against code that depends on getting the current time, the best approach is to use a proven mechanism for setting the current time explicitly. It is the safest and most reliable way of testing time critical code.

It is true that the code does not run in frozen time, however, a unit testing environment is artificial and we need to be able to control the environment. Freezing and manipulating the wall-clock time is a tried and true approach.

For example, the Armasec unit tests use this when dealing with token expiration logic: https://github.com/omnivector-solutions/armasec/blob/main/tests/test_token_manager.py#L68

Reply:
You have a fair point.

Looking at [freezegun's source code](https://github.com/spulec/freezegun/blob/ac4d92ec6d3545b113aba96dc444fd055219af9b/freezegun/api.py#L816) we can see they basically also mock the time with a different approach. I cannot see the difference that can lead us in edge cases. Can you elaborate a little bit more the edge cases you see? It could help me to develop new tests as well.

I think having it not frozen enables us to know if the code can run at any time, which indeed it is supposed to.

Reply:
Here is an interesting article about the problem and how a library helps:
https://adamj.eu/tech/2020/06/03/introducing-time-machine/

Freezing time is important because you don't know when a test will be run. Sometimes your logic makes assumption about events happening in the same day (what if the test runs exactly at midnight) or same hour (what happens if the test runs exactly at the hour). When you are creating tests that involve time, it's really important that you control the time parameter. That way, you can test that it works correctly in a normal situation and also covers the odd edge-cases.

This is best practice and an important tool to use when testing around timestamps.

Reply:
The article is indeed very interesting, but look at his point on why mock isn't good for mocking time:

[Using unittest.mock works, but it’s often inaccurate as each patcher can only mock a single function reference. This inaccuracy is exacerbated when mocking the current time, as there are many different functions that return the current time, in different formats. Due to the way Python’s imports work, it takes a lot of mocks to replace every instance of functions from datetime and time in a code path, and is sometimes impossible.](https://adamj.eu/tech/2020/06/03/introducing-time-machine/#:~:text=Using%20unittest.mock,is%20sometimes%20impossible.)

If you think precisely on this you shall realize that the current implementation I did has nothing against Mr. Johnson's thinking.

Also, I found very weird the fact these other packages for freezing the time change the C level of python. This definitely won't simulate the environment in which the code will run. What if I change the python interpreter then? What problems could we have? If you see, every approach has its downsides.

If we want some granularity here I'd prefer to generate the datetime randomly based on a seed.

I also want other folks' thinkings because I'm stricted convinced using this other packages isn't a good practice at all.

@julianaklulo @fschuch give your thoughts please

Reply:
I've been using `freezegun` since the first time I needed to manipulate the current time in a test, because it is the recommended approach for the reasons @dusktreader already mentioned above.

I agree that mocking the `datetime` directly is an odd way of doing this, and since we've been using `pytest-freezegun` in other projects, I'd suggest we set it as a standard across all projects.

Reply:
according to private conversations with @dusktreader, we agreed the current implementation is fine but he alerted with examples why freezegun or timemachine are good packages for handling timestamp manipulation during tests. Since we resolved this conversation in private I will resolve here as well.

Furthermore, you folks feel free to give your thoughts even if the conversation is resolved.